db/db_impl_secondary.cc

```diff
@@ -52,12 +62,200 @@ Status DBImplSecondary::Recover(
+      }
+    }
+  };
+  auto iter = log_readers_.find(log_number);
+  // make sure the log file is still present
+  if (iter == log_readers_.end() ||
+      iter->second->GetLogNumber() != log_number) {
+    // delete the obsolete log reader if log number mismatch
+    if (iter != log_readers_.end()) {
+      delete iter->second;
+      log_readers_.erase(iter);
+      log_reporters_.erase(log_number);
+      log_reader_status_.erase(log_number);
+    }
+    // initialize log reader from log_number
+    // TODO: min_log_number_to_keep_2pc check needed?
+    // Open the log file
+    std::string fname = LogFileName(immutable_db_options_.wal_dir, log_number);
+    ROCKS_LOG_INFO(immutable_db_options_.info_log,
+                   "Recovering log #%" PRIu64 " mode %d", log_number,
+                   static_cast<int>(immutable_db_options_.wal_recovery_mode));
+
+    std::unique_ptr<SequentialFileReader> file_reader;
+    {
+      std::unique_ptr<SequentialFile> file;
+      Status status = env_->NewSequentialFile(fname, &file,
+                                       env_->OptimizeForLogRead(env_options_));
+      if (!status.ok()) {
+        MaybeIgnoreError(&status);
+        *log_reader = nullptr;
+        return status;
+      }
+      file_reader.reset(new SequentialFileReader(std::move(file), fname));
+    }
+
+    // Create the log reader.
+    LogReporter* reporter = new LogReporter();
+    Status* status = new Status();
+    reporter->env = env_;
+    reporter->info_log = immutable_db_options_.info_log.get();
+    reporter->fname = fname.c_str();
+    if (!immutable_db_options_.paranoid_checks ||
+        immutable_db_options_.wal_recovery_mode ==
+            WALRecoveryMode::kSkipAnyCorruptedRecords) {
+      reporter->status = nullptr;
+    } else {
+      reporter->status = status;
+    }
+    // We intentially make log::Reader do checksumming even if
+    // paranoid_checks==false so that corruptions cause entire commits
+    // to be skipped instead of propagating bad information (like overly
+    // large sequence numbers).
+    log::FragmentBufferedReader* reader = new log::FragmentBufferedReader(immutable_db_options_.info_log,
+        std::move(file_reader), reporter, true /*checksum*/, log_number);
+
+    log_readers_.insert({log_number, reader});
+    log_reporters_.insert({log_number, reporter});
+    log_reader_status_.insert({log_number, status});
+  }
+  iter = log_readers_.find(log_number);
+  assert(iter != log_readers_.end());
+  *log_reader = iter->second;
+  return Status::OK();
+}
+
+// After manifest recovery, replay WALs and refresh log_readers_ if necessary
+// REQUIRES: log_numbers are sorted in ascending order
+Status DBImplSecondary::RecoverLogFiles(
+    const std::vector<uint64_t>& log_numbers, SequenceNumber* next_sequence,
+    bool /*read_only*/) {
+  mutex_.AssertHeld();
+  Status status;
+  // for (auto cfd : *versions_->GetColumnFamilySet()) {
+  for (auto log_number : log_numbers) {
```

Comment:
Can we open the WAL in the beginning? If any of them fails to open, we just return.

Reply:
We still open one WAL, process it before opening the second one?

Reply:
Oh, misunderstood the original comment, so use two for-loops? first one calls `GetLogReader` and second one calls `ReadRecord`?

Reply:
Yes

Reply:
Added MaybeInitLogReader to initialize the log readers in the first iteration

Reply:
Have you pushed `MaybeInitLogReader` yet?